pkg/disk/partition_table.go

```diff
@@ -1144,3 +1112,286 @@ func mkESP(size uint64, ptType PartitionTableType) (Partition, error) {
+	// PartitionTableType must be either "dos" or "gpt". Defaults to "gpt".
+	PartitionTableType PartitionTableType
+
+	// BootMode determines the types of boot-related partitions that are
+	// automatically added, BIOS boot (legacy), ESP (UEFI), or both (hybrid).
+	// If none, no boot-related partitions are created.
+	BootMode platform.BootMode
+
+	// DefaultFSType determines the filesystem type for automatically created
+	// filesystems and custom mountpoints that don't specify a type.
+	// None is only valid if no partitions are created and all mountpoints
+	// partitions specify a type.
+	// The default type is also used for the automatically created /boot
+	// filesystem if it is a supported type for that fileystem. If it is not,
+	// xfs is used as a fallback.
+	DefaultFSType FSType
+
+	// RequiredMinSizes defines a map of minimum sizes for specific
+	// directories. These indirectly control the minimum sizes of partitions. A
+	// directory with a required size will set the minimum size of the
+	// partition with the mountpoint that contains the directory. Additional
+	// directory requirements are additive, meaning the minimum size for a
+	// mountpoint's partition is the sum of all the required directory sizes it
+	// will contain.
+	RequiredMinSizes map[string]uint64
+}
+
+// Returns the default filesystem type if the fstype is empty. If both are
+// empty/none, returns an error.
+func (options *CustomPartitionTableOptions) getfstype(fstype string) (string, error) {
+	if fstype != "" {
+		return fstype, nil
+	}
+
+	if options.DefaultFSType == FS_NONE {
+		return "", fmt.Errorf("no filesystem type defined and no default set")
+	}
+
+	return options.DefaultFSType.String(), nil
+}
+
+// NewCustomPartitionTable creates a partition table based almost entirely on the disk customizations from a blueprint.
+func NewCustomPartitionTable(customizations *blueprint.DiskCustomization, options *CustomPartitionTableOptions, rng *rand.Rand) (*PartitionTable, error) {
+	if options == nil {
+		// init options with defaults
+		options = &CustomPartitionTableOptions{
+			PartitionTableType: PT_GPT,
+		}
+	}
+
+	if customizations == nil {
+		customizations = &blueprint.DiskCustomization{}
+	}
+
+	errPrefix := "error generating partition table:"
+
+	// validate the partitioning customizations before using them
+	if err := customizations.Validate(); err != nil {
+		return nil, fmt.Errorf("%s %w", errPrefix, err)
+	}
+
+	pt := &PartitionTable{}
+
+	// TODO: Handle partition table type in customizations
+	switch options.PartitionTableType {
+	case PT_GPT, PT_DOS:
+		pt.Type = options.PartitionTableType
+	case PT_NONE:
+		// default to "gpt"
+		pt.Type = PT_GPT
+	default:
+		return nil, fmt.Errorf("%s invalid partition table type enum value: %d", errPrefix, options.PartitionTableType)
+	}
+
+	// TODO: switch to ensure ESP in case customizations already include it
+	if err := AddPartitionsForBootMode(pt, options.BootMode); err != nil {
```

Comment:
The function was added in a previous PR when we were breaking up the original one.  If it was unexported, it would be blocked by the linter as dead code.  We can make them private again now, but maybe they could be useful for building partition tables externally?  Maybe we should just make them private until we discover we need them.

Reply:
Will also be part of the followup.

Reply:
I think unexporting for now is fine, I did that in https://github.com/osbuild/images/pull/1062